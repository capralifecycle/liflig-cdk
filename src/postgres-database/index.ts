export { PostgresDatabase } from "./postgres-database"
