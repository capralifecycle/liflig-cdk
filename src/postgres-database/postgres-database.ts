import * as ec2 from "@aws-cdk/aws-ec2"
import * as rds from "@aws-cdk/aws-rds"
import * as sm from "@aws-cdk/aws-secretsmanager"
import * as cdk from "@aws-cdk/core"

interface Props extends cdk.StackProps {
  vpc: ec2.IVpc
  /**
   * @default master
   */
  masterUsername?: string
  /**
   * @default app
   */
  databaseName?: string
  /**
   * @default 25
   */
  allocatedStorageGb?: number
  instanceType: ec2.InstanceType
  instanceIdentifier: string
  /**
   * Must not be removed once it has been set, as changing this
   * results in a new DB instance being created.
   *
   * Also, remember to give database a new name when changing this prop, or else
   * the new instance name will crash with the existing instance.
   */
  snapshotIdentifier?: string
  /**
   * @default false
   */
  usePublicSubnets?: boolean
}

export class PostgresDatabase extends cdk.Construct {
  public readonly secret: sm.ISecret
  public readonly connections: ec2.Connections

  constructor(scope: cdk.Construct, id: string, props: Props) {
    super(scope, id)

    const masterUsername = props.masterUsername ?? "master"
    const databaseName = props.databaseName ?? "app"

    const dbSecret = new rds.DatabaseSecret(this, "DbSecret", {
      username: masterUsername,
    })

    const dbOptions: rds.DatabaseInstanceSourceProps = {
      engine: rds.DatabaseInstanceEngine.postgres({
        version: rds.PostgresEngineVersion.VER_12,
      }),
      allowMajorVersionUpgrade: true,
      instanceIdentifier: props.instanceIdentifier,
      instanceType: props.instanceType,
      vpc: props.vpc,
      vpcSubnets: props.usePublicSubnets
        ? {
            subnetType: ec2.SubnetType.PUBLIC,
          }
        : undefined,
      multiAz: true,
      // We default to 25 GiB storage instead of 100 GiB
      // if we do not specify.
      allocatedStorage: props.allocatedStorageGb ?? 25,
      // We specify maximum backup retention.
      backupRetention: cdk.Duration.days(35),
    }

    const db = props.snapshotIdentifier
      ? new rds.DatabaseInstanceFromSnapshot(this, "Db", {
          ...dbOptions,
          snapshotIdentifier: props.snapshotIdentifier,
          credentials: rds.SnapshotCredentials.fromSecret(dbSecret),
        })
      : new rds.DatabaseInstance(this, "Db", {
          ...dbOptions,
          databaseName,
          credentials: rds.Credentials.fromSecret(dbSecret),
          storageEncrypted: true,
        })

    this.secret = db.secret!
    this.connections = db.connections

    if (props.usePublicSubnets) {
      // Override due to being in public subnet.
      // It will default to being public accessible which we do not want.
      ;(db.node.defaultChild as rds.CfnDBInstance).publiclyAccessible = false
    }
  }
}
