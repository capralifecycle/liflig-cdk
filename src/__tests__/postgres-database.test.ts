import "@aws-cdk/assert/jest"
import { App, Stack } from "@aws-cdk/core"
import * as ec2 from "@aws-cdk/aws-ec2"
import "jest-cdk-snapshot"
import { PostgresDatabase } from "../postgres-database"

test("create postgres database", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")

  const vpc = new ec2.Vpc(stack, "Vpc", {
    subnetConfiguration: [
      {
        cidrMask: 19,
        name: "Public",
        subnetType: ec2.SubnetType.PUBLIC,
      },
    ],
  })

  new PostgresDatabase(stack, "Database", {
    vpc: vpc,
    instanceType: ec2.InstanceType.of(
      ec2.InstanceClass.BURSTABLE3,
      ec2.InstanceSize.MICRO,
    ),
    instanceIdentifier: "example-database-v1",
    snapshotIdentifier: undefined,
    usePublicSubnets: true,
  })

  expect(stack).toMatchCdkSnapshot()
})
