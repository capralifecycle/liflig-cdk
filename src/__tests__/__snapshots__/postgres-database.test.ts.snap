// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create postgres database 1`] = `
Object {
  "Resources": Object {
    "DatabaseDbE7C02057": Object {
      "DeletionPolicy": "Snapshot",
      "Properties": Object {
        "AllocatedStorage": "25",
        "AllowMajorVersionUpgrade": true,
        "BackupRetentionPeriod": 35,
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.t3.micro",
        "DBInstanceIdentifier": "example-database-v1",
        "DBName": "app",
        "DBSubnetGroupName": Object {
          "Ref": "DatabaseDbSubnetGroupFB62376A",
        },
        "Engine": "postgres",
        "EngineVersion": "12",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DatabaseDbSecret1098DC6E",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DatabaseDbSecret1098DC6E",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "MultiAZ": true,
        "PubliclyAccessible": false,
        "StorageEncrypted": true,
        "StorageType": "gp2",
        "VPCSecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "DatabaseDbSecurityGroupE6612FBB",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Snapshot",
    },
    "DatabaseDbSecret1098DC6E": Object {
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Generated by the CDK for stack: ",
              Object {
                "Ref": "AWS::StackName",
              },
            ],
          ],
        },
        "GenerateSecretString": Object {
          "ExcludeCharacters": " %+~\`#$&*()|[]{}:;<>?!'/@\\"\\\\",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{\\"username\\":\\"master\\"}",
        },
      },
      "Type": "AWS::SecretsManager::Secret",
    },
    "DatabaseDbSecretAttachment34026B56": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "DatabaseDbSecret1098DC6E",
        },
        "TargetId": Object {
          "Ref": "DatabaseDbE7C02057",
        },
        "TargetType": "AWS::RDS::DBInstance",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "DatabaseDbSecurityGroupE6612FBB": Object {
      "Properties": Object {
        "GroupDescription": "Security group for Db database",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DatabaseDbSubnetGroupFB62376A": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnet group for Db database",
        "SubnetIds": Array [
          Object {
            "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
          },
          Object {
            "Ref": "VpcPublicSubnet2Subnet691E08A3",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "Vpc8378EB38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcIGWD7BA715C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcPublicSubnet1DefaultRoute3DA9E72A": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet1RouteTable6C95E38E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet1RouteTableAssociation97140677": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet1Subnet5C2D37C4": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcPublicSubnet2DefaultRoute97F91067": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet2RouteTable94F7E489": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet2RouteTableAssociationDD5762D8": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet2Subnet691E08A3",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet2Subnet691E08A3": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.32.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Stack/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcVPCGWBF912B6E": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
